### DISCUSSION TITLE
Allow dynamic authorization header in SBT resolvers

### DISCUSSION THREAD

**Discussion by fiadliel** [sbt/sbt#7003] [Ideas]
SBT used to allow customisation of HTTP calls by using `CustomHttp.okhttpClientBuilder`. Before SBT 1.7, I had written a plugin which would create an Interceptor for a custom OkHttpClient builder to make changes on every HTTP request, in order to dynamically change the headers (I wanted to add a bearer token in an Authorization header). The header also needs to be sent even without a server auth challenge.
i.e. I want to always add a header like `Authorization: Bearer blahblahblah` where the value may change at arbitrary intervals, and code needs to be run to find out the current value.

After 1.7, it seems that such functionality is no longer available.

It is possible in Coursier because `csrConfiguration` is a task, so we can change the resolvers and associated authentication, on each request; Coursier also allows arbitrary `Authentication` headers to be specified for a resolver. But even if plumbed into Coursier, support is still needed in SBT for publishing.

Concrete use case: different people have been trying different mechanisms to plumb support for Google Artifact Registry into SBT, including registering custom protocol handlers, setting up different class loaders, and so forth. But the fundamental problem is so much simpler, we just need to add the right Authorization header, and otherwise existing SBT support works fine. This might also work for a lot of other systems in the future, since static authentication passwords are becoming an anti-pattern for security, especially in automated environments.

**Comment by jivanic-demystdata** [comment]
@fiadliel Did you find a solution? I need to publish to Buildkite Packages and they're using a (fake) "Bearer" token mechanism where we need to pass `Bearer <API key>`
<img width="766" alt="image" src="https://github.com/user-attachments/assets/d12e83bc-a03f-4e40-8723-de968f58409c">

(See their doc: https://buildkite.com/docs/packages/maven#publish-a-package)

**Reply by fiadliel** [reply]
@jivanic-demystdata I didn't get any traction on the request.

You don't actually need a dynamic header, just a static one in this case. This is also a stubbornly missing feature in SBT. But looking through https://github.com/sbt/sbt/issues/4382, I see that https://github.com/arktekk/sbt-aether-deploy (deploy using Maven) has support for static custom headers, so you may be able to get this to work.

**Reply by jivanic-demystdata** [reply]
Thanks for your help @fiadliel! ❤️
I'll have a look

**Reply by jivanic-demystdata** [reply]
FYI @fiadliel, in the end, I forked an sbt plugin, made it work and published it to Maven Central: https://github.com/guizmaii-opensource/sbt-publish-with-headers

I don't know if it can work for your needs
